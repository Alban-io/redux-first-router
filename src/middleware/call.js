export default (name, { hasReturn, prev } = {}) => async (req, next) => {
  hasReturn = true

  const route = prev === 'prev' ? req.prevRoute : req.route
  const routeCb = route[name] || noop
  const globalCb = req.options[name] || noop

  const [a, b] = await Promise.all([routeCb(req), globalCb(req)])
  await next()

  return hasReturn ? a || b : undefined
}

const noop = function () {}
