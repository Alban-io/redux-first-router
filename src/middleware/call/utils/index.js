export { default as complete } from './complete'
export { default as enhanceRoutes } from './enhanceRoutes'
export { default as isAutoDispatch } from './isAutoDispatch'
export { default as isGlobalSkip } from './isGlobalSkip'
export { default as onError } from './onError'
export { default as shouldCall } from './shouldCall'
export { default as isFalse } from './isFalse'
