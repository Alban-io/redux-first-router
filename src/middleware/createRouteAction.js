import pathToAction from '../pure-utils/pathToAction'
import actionToPath from '../pure-utils/actionToPath'
import isRedirect from '../pure-utils/isRedirect'
import { NOT_FOUND } from '../index'

export default async (req, next) => {
  const {
    history,
    routesMap,
    options: { querySerializer: serializer, notFoundPath },
    getLocationState,
    nextHistory,
    action
  } = req

  const state = getLocationState()
  const { hasSSR, routesMap: r, prev: p, entries: e, ...prev } = state

  try {
    if (nextHistory) {
      const { url } = nextHistory.location
      const action = pathToAction(url, routesMap, serializer)

      req.action = nestAction(url, action, prev, nextHistory)
      req.route = routesMap[action.type]
    }
    else if (action && action.type !== NOT_FOUND) {
      const url = actionToPath(action, routesMap, serializer)
      const method = isRedirect(action) ? 'redirect' : 'push'
      const { nextHistory, commit } = history[method](url, {}, false)

      req.action = nestAction(url, action, prev, nextHistory)
      req.nextHistory = nextHistory
      req.commit = commit
    }
  }
  catch (e) {
    req.action = nestAction(
      notFoundPath || prev.pathname || '/',
      { ...action, type: NOT_FOUND, payload: { ...action.payload } },
      prev,
      nextHistory
    )
  }

  await next()
  return req.action
}

const nestAction = (pathname, action, prev, nextHistory) => {
  const { kind } = nextHistory
  const { type, payload = {}, meta = {} } = action
  const query = action.query || meta.query || payload.query
  const parts = pathname.split('?')
  const search = parts[1]

  return {
    kind,
    ...action,
    ...(action.query && { query }),
    type,
    payload,
    meta: {
      ...meta,
      ...(meta.query && { query }),
      location: {
        current: {
          pathname: parts[0],
          type,
          payload,
          kind,
          ...(query && { query, search })
        },
        kind,
        prev,
        history: nextHistory
      }
    }
  }
}
