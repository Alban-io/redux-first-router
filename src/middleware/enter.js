export default (api) => async (req, next) => {
  const ret = req.commitDispatch(req.action)

  req.commitHistory()
  // req.ctx.busy = false
  req.tmp.committed = true

  await next()
  return ret
}
