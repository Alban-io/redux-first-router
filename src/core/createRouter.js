// @flow
import type { RoutesMapInput, Options, Store, Dispatch } from '../flow-types'

import composePromise from './composePromise'
import createLocationReducer from './createLocationReducer'
import createSmartHistory from '../history'

import {
  createSelector,
  formatRoutes,
  shouldTransition,
  createRequest
} from '../utils'

import {
  serverRedirect,
  addRoutes,
  pathlessThunk,
  anonymousThunk,
  transformAction,
  call,
  enter,
  changePageTitle
} from '../middleware'

import { onError as defaultOnError } from '../middleware/call/utils'

export default (
  routesInput: RoutesMapInput = {},
  options: Options = {},
  middlewares: Array<Function> = [
    serverRedirect,     // short-circuiting middleware
    addRoutes,
    pathlessThunk,
    anonymousThunk,
    transformAction,      // pipeline starts here
    call('beforeLeave', { prev: true }),
    call('beforeEnter'),
    enter,
    changePageTitle,
    call('onLeave', { prev: true }),
    call('onEnter'),
    call('thunk', { cache: true }),
    call('onComplete')
  ]
) => {
  const {
    location,
    title,
    formatRoute,
    createHistory = createSmartHistory,
    createReducer = createLocationReducer,
    compose = composePromise,
    onError
  } = options

  options.shouldTransition = options.shouldTransition || shouldTransition
  options.createRequest = options.createRequest || createRequest
  options.onError = typeof onError !== 'undefined' ? onError : defaultOnError

  const routes = formatRoutes(routesInput, formatRoute)
  const selectLocationState = createSelector('location', location)
  const selectTitleState = createSelector('title', title)
  const history = createHistory(options)
  const reducer = createReducer(routes, history.firstRoute.nextHistory, options)
  const api = { routes, history, options }

  const middleware = (store: Store) => {
    const getTitle = () => selectTitleState(store.getState() || {})
    const getLocation = (s) => selectLocationState(s || store.getState() || {})

    Object.assign(api, { store, getTitle, getLocation })

    const nextPromise = compose(middlewares, api, true)
    const { shouldTransition, createRequest } = options
    const onError = call('onError')(api)
    let tmp = {}

    history.listen(store.dispatch)
    store.getState.rudy = api // make rudy available via `context` (see <Link />)

    return (next: Dispatch) => (action: Object) => {
      if (!shouldTransition(action, api)) return next(action)

      const req = createRequest(action, api, tmp, next)

      return nextPromise(req) // start middleware pipeline
        .catch(error => {
          req.error = error
          req.errorType = `${req.action.type}_ERROR`
          return onError(req)
        })
        .then(res => {
          tmp = {}
          req.completed = true
          return res
        })
    }
  }

  return {
    firstRoute: () => history.firstRoute,
    middleware,
    reducer,
    rudy: api
  }
}

