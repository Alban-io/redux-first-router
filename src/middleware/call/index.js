import { enhanceRoutes, shouldCall, createCache } from './utils'
import { noOp, isAction } from '../../utils'

export default (name, config = {}) => (api) => {
  const { cache, prev, skipOpts = false } = config

  enhanceRoutes(name, api.routes)
  api.options.callbacks = api.options.callbacks || []
  api.options.callbacks.push(name)

  api.options.shouldCall = api.options.shouldCall || shouldCall

  if (cache) {
    api.clearCache = createCache(api, name)
  }

  return (req, next = noOp) => {
    const route = prev ? req.prevRoute : req.route
    const execute = req.options.shouldCall(name, route, req, config)

    if (!execute) return next()

    const r = (execute.route && route[name]) || noOp
    const o = (execute.options && !skipOpts && req.options[name]) || noOp

    req._dispatched = false // `dispatch` used by callbacks will set this to `true` (see utils/createDispatch.js)

    return Promise.all([r(req), o(req)]).then(([r, o]) => {
      if (isFalse(r, o)) return false
      const res = r || o

      if (res && !req._dispatched && isAutoDispatch(route, req.options)) { // if no dispatch was detected, and a result was returned, dispatch it automatically
        const action = isAction(res) ? res : { payload: res }
        action.type = action.type || `${req.action.type}_COMPLETE`

        return Promise.resolve(req.dispatch(action))
          .then(complete(next))
      }

      return complete(next)(res)
    })
  }
}

const isFalse = (r, o) => r === false || o === false

const complete = (next) => (res) => next().then(() => res)

const isAutoDispatch = (route, options) =>
  route.autoDispatch !== undefined
    ? route.autoDispatch
    : options.autoDispatch === undefined ? true : options.autoDispatch
