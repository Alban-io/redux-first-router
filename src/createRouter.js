// @flow
import composePromise from './composePromise'
import createSmartHistory from './smart-history'
import createLocationReducer from './createLocationReducer'

import createSelector from './utils/createSelector'
import createDispatch from './utils/createDispatch'
import formatRoutes from './utils/formatRoutes'
import shouldTrans from './utils/shouldTransition'

import serverRedirect from './middleware/serverRedirect'
import addRoutes from './middleware/addRoutes'
import pathlessThunk from './middleware/pathlessThunk'
import plainActionRoute from './middleware/plainActionRoute'
import anonymousThunk from './middleware/anonymousThunk'
import createRouteAction from './middleware/createRouteAction'
import call from './middleware/call'
import enter from './middleware/enter'
import changePageTitle from './middleware/changePageTitle'

import type { RoutesMapInput, Options, Store, Dispatch } from './flow-types'
import { ERROR } from './index'

export default (
  routesInput: RoutesMapInput = {},
  options: Options = {},
  middlewares: Array<Function> = [
    serverRedirect,     // short-circuiting middleware
    addRoutes,
    pathlessThunk,
    plainActionRoute,
    anonymousThunk,
    createRouteAction,  // pipeline starts here
    call('beforeLeave', { prev: true }),
    call('beforeEnter'),
    enter,
    changePageTitle,
    call('onLeave', { prev: true }),
    call('onEnter'),
    call('thunk'),
    call('onComplete')
  ]
) => {
  const {
    location,
    title,
    createHistory = createSmartHistory,
    createReducer = createLocationReducer,
    shouldTransition = shouldTrans
  } = options

  options.shouldTransition = shouldTransition

  const routes = formatRoutes(routesInput)
  const selectLocationState = createSelector('location', location)
  const selectTitleState = createSelector('title', title)
  const history = createHistory(options)
  const reducer = createReducer(routes, history, options)

  const middleware = (store: Store) => {
    const getTitle = () => selectTitleState(store.getState() || {})
    const locationState = () => selectLocationState(store.getState() || {})
    const ctx = {}
    const tmp = {}
    const api = { store, history, routes, options, locationState, ctx, tmp }
    const nextPromise = composePromise(middlewares, api)
    const shouldTransition = options.shouldTransition
    const onError = call('onError')(api)

    history.listen(store.dispatch)
    store.getState.rudy = api // make rudy available via `context` (see <Link />)

    return (next: Dispatch) => (action: Object) => {
      if (!shouldTransition(action, api)) return next(action)

      const req = {
        ...options.extra,
        ...api,
        getTitle,
        state: store.getState(),
        getState: store.getState,
        location: locationState(),
        dispatch: createDispatch(() => req),
        action: !action.nextHistory ? action : null,
        prevRoute: routes[locationState().type],
        route: routes[action.type],
        nextHistory: action.nextHistory || null,
        commitHistory: action.commit || null,
        commitDispatch: next,
        completed: false,
        error: null
      }

      return nextPromise(req) // start middleware pipeline
        .catch(error => {
          req.error = error
          req.errorType = req.route ? `${req.action.type}_ERROR` : ERROR
          console.log('ERROR!!', error.stack.replace(new RegExp('/Users/jamesgillmore/.vscode/extensions/WallabyJs.wallaby-vscode-1.0.64/projects/2c9e7f1cfb906e5d/instrumented', 'g'), ''))
          return onError(req)
        })
        .then(res => {
          for (const key in tmp) delete tmp[key]
          req.completed = true
          return res
        })
    }
  }

  _history = history

  return {
    middleware,
    reducer,
    history,
    firstRoute: () => history.firstRoute
  }
}

let _history
export const history = () => _history
