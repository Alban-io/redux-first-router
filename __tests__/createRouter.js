import { applyMiddleware, createStore, compose, combineReducers } from 'redux'
import reduxThunk from 'redux-thunk'
import createRouter from '../src/createRouter'

import call from '../src/middleware/call'
import enter from '../src/middleware/enter'
import createRouteAction from '../src/middleware/createRouteAction'

const setup = (path = '/first', options = {}) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second',
    THIRD: '/third'
  }

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const routerMiddlewares = [
    createRouteAction,
    enter
  ]

  const { enhancer, reducer, firstRoute, history } = createRouter(
    routesMap,
    options,
    routerMiddlewares
  )

  const middlewares = applyMiddleware(reduxThunk)
  const enhancers = compose(enhancer, middlewares)
  const title = (state = {}, action = {}) => action.type
  const rootReducer = combineReducers({ title, location: reducer })
  const store = createStore(rootReducer, enhancers)

  return {
    store,
    firstRoute,
    history
  }
}

test('store.dispatch', async () => {
  const { store, firstRoute, history } = setup()
  const action = firstRoute()

  let res = await store.dispatch(action)
  expect(store.getState().location.type).toEqual('FIRST')

  res = await store.dispatch({ type: 'SECOND' })
  expect(store.getState().location.type).toEqual('SECOND')
  expect(res.type).toEqual('SECOND')
})


test('history.push', async () => {
  const { store, firstRoute, history } = setup()
  const action = firstRoute()

  let res = await store.dispatch(action)
  expect(store.getState().location.type).toEqual('FIRST')

  res = await history.push('/second')
  expect(store.getState().location.type).toEqual('SECOND')
  expect(res.type).toEqual('SECOND')
})


// const routerMiddlewares = [
//   async (bag, next) => {
//     console.log('MIDDLEWARE 1!', bag)
//     const res = bag.dispatch({ type: 'BLA' })
//     await next()
//     return res
//   }
// ]
