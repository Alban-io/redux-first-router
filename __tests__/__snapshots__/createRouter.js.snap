// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createActions 1`] = `
Object {
  "FIRST": "SCENE/FIRST",
  "FIRST_COMPLETE": "SCENE/FIRST_COMPLETE",
  "FIRST_ERROR": "SCENE/FIRST_ERROR",
  "FOURTH": "SCENE/FOURTH",
  "FOURTH_COMPLETE": "SCENE/FOURTH_COMPLETE",
  "FOURTH_ERROR": "SCENE/FOURTH_ERROR",
  "NOT_FOUND": "SCENE/@@redux-first-router/NOT_FOUND",
  "NOT_FOUND_COMPLETE": "SCENE/@@redux-first-router/NOT_FOUND_COMPLETE",
  "NOT_FOUND_ERROR": "SCENE/@@redux-first-router/NOT_FOUND_ERROR",
  "PLAIN": "SCENE/PLAIN",
  "PLAIN_COMPLETE": "SCENE/PLAIN_COMPLETE",
  "PLAIN_ERROR": "SCENE/PLAIN_ERROR",
  "SECOND": "SCENE/SECOND",
  "SECOND_COMPLETE": "SCENE/SECOND_COMPLETE",
  "SECOND_ERROR": "SCENE/SECOND_ERROR",
  "THIRD": "SCENE/THIRD",
  "THIRD_COMPLETE": "SCENE/THIRD_COMPLETE",
  "THIRD_ERROR": "SCENE/THIRD_ERROR",
}
`;

exports[`createActions 2`] = `
Object {
  "first": [Function],
  "firstComplete": [Function],
  "firstError": [Function],
  "fourth": [Function],
  "fourthComplete": [Function],
  "fourthError": [Function],
  "notFound": undefined,
  "notFoundComplete": undefined,
  "notFoundError": undefined,
  "plain": [Function],
  "plainComplete": [Function],
  "plainError": [Function],
  "reduxFirstRouterNotFound": [Function],
  "reduxFirstRouterNotFoundComplete": [Function],
  "reduxFirstRouterNotFoundError": [Function],
  "second": [Function],
  "secondComplete": [Function],
  "secondError": [Function],
  "third": [Function],
  "thirdComplete": [Function],
  "thirdError": [Function],
}
`;

exports[`createActions 3`] = `
"const { types, actions } = createActions(routes)

const { 
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
} = types

const { 
	first,
	firstComplete,
	firstError,
	second,
	secondComplete,
	secondError,
	third,
	thirdComplete,
	thirdError,
	fourth,
	fourthComplete,
	fourthError,
	plain,
	plainComplete,
	plainError,
	reduxFirstRouterNotFound,
	reduxFirstRouterNotFoundComplete,
	reduxFirstRouterNotFoundError,
	notFound,
	notFoundComplete,
	notFoundError
} = actions

export {
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
}

export {
	first,
	firstComplete,
	firstError,
	second,
	secondComplete,
	secondError,
	third,
	thirdComplete,
	thirdError,
	fourth,
	fourthComplete,
	fourthError,
	plain,
	plainComplete,
	plainError,
	reduxFirstRouterNotFound,
	reduxFirstRouterNotFoundComplete,
	reduxFirstRouterNotFoundError,
	notFound,
	notFoundComplete,
	notFoundError
}"
`;
